Upgrade to Ubuntu Focal 20.04 LTS

nfsec opened the pull request:
- Upgrade Docker images to Ubuntu Focal 20.04 LTS.
- Standardization of pip install command.
- apt command optimization.

nfsec committed:
Update Dockerfile

nfsec committed:
Update Dockerfile

nfsec committed:
Update Dockerfile

nfsec committed:
Update Dockerfile

nfsec committed:
Update Dockerfile

Arusekk commented on extra/docker/develop/Dockerfile:
```diff
@@ -51,20 +50,17 @@ RUN sudo apt-get install -y \
-	zsh
-
+	zsh \
 # Misc useful things when developing
-RUN sudo apt-get install -y  \
 	curl \
```
Does this work with the comment between escaped newlines?

nfsec replied:
Yes.

Arusekk commented on extra/docker/develop/Dockerfile:
```diff
@@ -18,20 +17,20 @@ ENV PATH="/home/pwntools/.local/bin:${PATH}"
- && python  -m pip install --upgrade --editable pwntools \
  && python3 -m pip install --upgrade --editable pwntools \
  && PWNLIB_NOTERM=1 pwn version
 
 # Requirements for running the tests
-RUN python  -m pip install --upgrade --requirement pwntools/docs/requirements.txt \
- && python3 -m pip install --upgrade --requirement pwntools/docs/requirements.txt
+RUN python3 -m pip install --upgrade --requirement pwntools/docs/requirements.txt
 
 # Python niceties for debugging
-RUN python  -m pip install -U ipython ipdb \
- && python3 -m pip install -U ipython ipdb
+RUN python3 -m pip install -U ipython ipdb
```
While at it, could these `pip install`s be moved after the `apt install`s? It looks like we might install several packages twice. The dependencies should be in a single RUN call preferably, too.  We should also remove the apt cache after `apt install` to take advantage of Docker layer caching (here and in base).

nfsec replied:
OK

peace-maker commented:
This removes Python 2 from the images which breaks the local testing setup in `travis/docker`. It'd be great to keep python2 available in the image as long as pwntools still supports it.

Arusekk commented on extra/docker/base/Dockerfile:
```diff
@@ -3,26 +3,23 @@
 ############################################################
 
-FROM ubuntu:bionic
+FROM ubuntu:focal
```
We can go even further now.

```suggestion
FROM ubuntu:jammy
```

Arusekk commented on extra/docker/base/Dockerfile:
```diff
@@ -3,26 +3,23 @@
-        python \
-        python-pip \
-        python-dev \
         python3 \
```
We still need to support python 2 for testing.
```suggestion
        python2.7 \
        python2.7-dev \
        python3 \
```

Arusekk commented on extra/docker/base/Dockerfile:
```diff
@@ -35,14 +32,12 @@ RUN apt-get update \
-    && pip3 install --upgrade pip \
+    && locale-gen en_US.UTF-8 \
+    && update-locale LANG=en_US.UTF-8 \
+    && python3 -m pip install --upgrade pip \
```
```suggestion
    && python2.7 -m ensurepip \
    && python2.7 -m pip install --upgrade pip \
    && python2.7 -m pip install --upgrade pwntools \
    && python3 -m pip install --upgrade pip \
```

Arusekk commented on extra/docker/beta/Dockerfile:
```diff
@@ -1,7 +1,6 @@
 USER root
-RUN pip install --upgrade git+https://github.com/Gallopsled/pwntools@beta
-RUN pip3 install --upgrade git+https://github.com/Gallopsled/pwntools@beta
+RUN python3 -m pip install --upgrade git+https://github.com/Gallopsled/pwntools@beta
```
```suggestion
RUN python2.7 -m pip install --upgrade git+https://github.com/Gallopsled/pwntools@beta \
    && python3 -m pip install --upgrade git+https://github.com/Gallopsled/pwntools@beta
```

Arusekk commented on extra/docker/dev/Dockerfile:
```diff
@@ -1,7 +1,6 @@
 USER root
-RUN pip install --upgrade git+https://github.com/Gallopsled/pwntools@dev
-RUN pip3 install --upgrade git+https://github.com/Gallopsled/pwntools@dev
+RUN python3 -m pip install --upgrade git+https://github.com/Gallopsled/pwntools@dev
```
```suggestion
RUN python2.7 -m pip install --upgrade git+https://github.com/Gallopsled/pwntools@dev \
    && python3 -m pip install --upgrade git+https://github.com/Gallopsled/pwntools@dev
```

Arusekk commented on extra/docker/develop/Dockerfile:
```diff
@@ -5,8 +5,7 @@ ENV HISTFILE=/home/pwntools/.history
 USER root
-RUN python -m pip uninstall -q -y pwntools \
- && python3 -m pip uninstall -q -y pwntools
+RUN python3 -m pip uninstall -q -y pwntools
```
```suggestion
RUN python2.7 -m pip uninstall -q -y pwntools \
    && python3 -m pip uninstall -q -y pwntools
```

Arusekk commented on extra/docker/develop/Dockerfile:
```diff
@@ -18,20 +17,20 @@ ENV PATH="/home/pwntools/.local/bin:${PATH}"
 # Install Pwntools to the home directory, make it an editable install
 RUN git clone https://github.com/Gallopsled/pwntools \
- && python  -m pip install --upgrade --editable pwntools \
  && python3 -m pip install --upgrade --editable pwntools \
```
```suggestion
 && python2.7 -m pip install --upgrade --editable pwntools \
 && python3 -m pip install --upgrade --editable pwntools \
```

Arusekk commented on extra/docker/develop/Dockerfile:
```diff
@@ -18,20 +17,20 @@ ENV PATH="/home/pwntools/.local/bin:${PATH}"
 # Requirements for running the tests
-RUN python  -m pip install --upgrade --requirement pwntools/docs/requirements.txt \
- && python3 -m pip install --upgrade --requirement pwntools/docs/requirements.txt
+RUN python3 -m pip install --upgrade --requirement pwntools/docs/requirements.txt
```
```suggestion
RUN python2.7 -m pip install --upgrade --requirement pwntools/docs/requirements.txt \
    && python3 -m pip install --upgrade --requirement pwntools/docs/requirements.txt
```

Arusekk commented on extra/docker/develop/Dockerfile:
```diff
@@ -18,20 +17,20 @@ ENV PATH="/home/pwntools/.local/bin:${PATH}"
 # Python niceties for debugging
-RUN python  -m pip install -U ipython ipdb \
- && python3 -m pip install -U ipython ipdb
+RUN python3 -m pip install -U ipython ipdb
```
```suggestion
RUN python2.7 -m pip install -U ipython ipdb \
    && python3 -m pip install -U ipython ipdb
```

Arusekk commented on extra/docker/stable/Dockerfile:
```diff
@@ -1,7 +1,6 @@
 USER root
-RUN pip install --upgrade git+https://github.com/Gallopsled/pwntools@stable
-RUN pip3 install --upgrade git+https://github.com/Gallopsled/pwntools@stable
+RUN python3 -m pip install --upgrade git+https://github.com/Gallopsled/pwntools@stable
```
```suggestion
RUN python2.7 -m pip install --upgrade git+https://github.com/Gallopsled/pwntools@stable \
    && python3 -m pip install --upgrade git+https://github.com/Gallopsled/pwntools@stable
```

Arusekk committed:
Move to Ubuntu Jammy 22.04 and keep python2.7

Arusekk committed:
Merge branch 'dev' into dev

Arusekk committed:
Apparently Debian intentionally broke ensurepip

Dear packagers,
I know that you do not like anything interfering with package managers
BUT ensurepip was invented in order to be the universal way to install
pip when NOTHING ELSE WORKS.  So I am quite upset that you broke it.
Thanks a lot anyway.
